src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py

```diff
@@ -49,7 +49,6 @@ class GetCompletedBlocks(DBFormatter):
                       dbsbuffer_dataset_subscription.site,
                       dbsbuffer_workflow.name,
                       dbsbuffer_block.create_time
-             HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
```

Comment:
That was useful in [GetDeletableBlocks.py](https://github.com/dmwm/WMCore/blob/master/src/python/WMComponent/RucioInjector/Database/MySQL/GetDeletableBlocks.py) because that query was not grouping by workflow name. Here, if you exclude records with workflows that are not complete, you will still have other records with other workflows associated to the block, so the block will still end up in the results, even with uncompleted associated workflows.
 
Even worst, as we are excluding the uncompleted workflows from the query, later we are not taking thos uncompleted workflows into account when we compare against the `deletableWfsDict` list

Reply:
>  because that query was not grouping by workflow name
That is correct. 
I have double checked now, that introducing the `dbsbuffer_workflow.name` in the `SELECT`  and `GROUP BY` conditions splits the grouping further down, creating  multiple groups per block - one for every associated workflow. And then it does count  the completed workflows but they enter the  query every time a workflow completes..... so if we have 1 block with 3 associated PromptReco requests out of which only two are completed, we will have that block returned 2 times  in the output. Once the third Prompt Reco is complete the same block would enter the output result 3 times.

This makes the rest of the logic quite hard to follow indeed.

Reply:
And since it is quite subtle detail and difficult to understand and remember. Let me put some concrete examples:

Here is the output from the current SQL query with the counts  per group included in the output (filtered for a single block):
```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_workflow.name,
  8     dbsbuffer_block.create_time
  9  FROM dbsbuffer_dataset_subscription
 10  INNER JOIN dbsbuffer_dataset ON
 11     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 12  INNER JOIN dbsbuffer_block ON
 13     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 14  INNER JOIN dbsbuffer_file ON
 15     dbsbuffer_file.block_id = dbsbuffer_block.id
 16  INNER JOIN dbsbuffer_workflow ON
 17     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 18  INNER JOIN dbsbuffer_location ON
 19     dbsbuffer_location.id = dbsbuffer_block.location
 20  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 21  AND dbsbuffer_dataset_subscription.subscribed = 1
 22  AND dbsbuffer_block.status = 'Closed'
 23  AND dbsbuffer_block.deleted = 0
 24  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 25   GROUP BY dbsbuffer_block.blockname,
 26     dbsbuffer_location.pnn,
 27     dbsbuffer_dataset.path,
 28     dbsbuffer_dataset_subscription.site,
 29     dbsbuffer_workflow.name,
 30     dbsbuffer_block.create_time
 31  HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
 32  ORDER BY dbsbuffer_workflow.name;

  COUNT(*) BLOCKNAME														    PNN 	    PATH				     SITE		NAME					 CREATE_TIME
---------- ------------------------------------------------------------------------------------------------------------------------ --------------- ---------------------------------------- ------------------ ---------------------------------------- -----------
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357080_ZeroBias		  1660249466
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357080_ZeroBias		  1660249466

``` 

Now lets release the `HAVING COUNT(*) = SUM(dbsbuffer_workflow.competed)` constraint:

```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_workflow.name,
  8     dbsbuffer_block.create_time
  9  FROM dbsbuffer_dataset_subscription
 10  INNER JOIN dbsbuffer_dataset ON
 11     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 12  INNER JOIN dbsbuffer_block ON
 13     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 14  INNER JOIN dbsbuffer_file ON
 15     dbsbuffer_file.block_id = dbsbuffer_block.id
 16  INNER JOIN dbsbuffer_workflow ON
 17     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 18  INNER JOIN dbsbuffer_location ON
 19     dbsbuffer_location.id = dbsbuffer_block.location
 20  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 21  AND dbsbuffer_dataset_subscription.subscribed = 1
 22  AND dbsbuffer_block.status = 'Closed'
 23  AND dbsbuffer_block.deleted = 0
 24  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 25   GROUP BY dbsbuffer_block.blockname,
 26     dbsbuffer_location.pnn,
 27     dbsbuffer_dataset.path,
 28     dbsbuffer_dataset_subscription.site,
 29     dbsbuffer_workflow.name,
 30     dbsbuffer_block.create_time
 31  ORDER BY dbsbuffer_workflow.name;

  COUNT(*) BLOCKNAME														    PNN 	    PATH				     SITE		NAME					 CREATE_TIME
---------- ------------------------------------------------------------------------------------------------------------------------ --------------- ---------------------------------------- ------------------ ---------------------------------------- -----------
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357080_ZeroBias		  1660249466
	82 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357080_ZeroBias		  1660249466
	77 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357081_ZeroBias		  1660249466
	77 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357081_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357096_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357096_ZeroBias		  1660249466
	 8 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357098_ZeroBias		  1660249466
	 8 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357098_ZeroBias		  1660249466
	 7 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357100_ZeroBias		  1660249466
	 7 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357100_ZeroBias		  1660249466
	13 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357101_ZeroBias		  1660249466

	13 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357101_ZeroBias		  1660249466
	 9 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357102_ZeroBias		  1660249466
	 9 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357102_ZeroBias		  1660249466
	 1 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357106_ZeroBias		  1660249466
	 1 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357106_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	PromptReco_Run357112_ZeroBias		  1660249466
	 2 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	PromptReco_Run357112_ZeroBias		  1660249466

18 rows selected.

```

It is obvious that  this block would continue entering the output of the original query every time a PromptReco completes until the result of the original query does not give exactly 18 records as the one above.

And indeed the expected behavior is achieved once we have the field `dbsbuffer_workflow.name` , which is further splitting the groups, removed.

Here is the query without the extra splitting and without the constraint for all the associated workflows to be completed as before:
```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_block.create_time
  8  FROM dbsbuffer_dataset_subscription
  9  INNER JOIN dbsbuffer_dataset ON
 10     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 11  INNER JOIN dbsbuffer_block ON
 12     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 13  INNER JOIN dbsbuffer_file ON
 14     dbsbuffer_file.block_id = dbsbuffer_block.id
 15  INNER JOIN dbsbuffer_workflow ON
 16     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 17  INNER JOIN dbsbuffer_location ON
 18     dbsbuffer_location.id = dbsbuffer_block.location
 19  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 20  AND dbsbuffer_dataset_subscription.subscribed = 1
 21  AND dbsbuffer_block.status = 'Closed'
 22  AND dbsbuffer_block.deleted = 0
 23  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 24   GROUP BY dbsbuffer_block.blockname,
 25     dbsbuffer_location.pnn,
 26     dbsbuffer_dataset.path,
 27     dbsbuffer_dataset_subscription.site,
 28     dbsbuffer_block.create_time ;

  COUNT(*) BLOCKNAME														    PNN 	    PATH				     SITE		CREATE_TIME
---------- ------------------------------------------------------------------------------------------------------------------------ --------------- ---------------------------------------- ------------------ -----------
       201 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_Disk	 1660249466
       201 /ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00						    T0_CH_CERN_Disk /ZeroBias/Run2022C-PromptReco-v1/AOD     T1_US_FNAL_MSS	 1660249466

```

And here is the result if we apply the constraint for workflow completion.  
```
SQL> SELECT
  2     count(*),
  3     dbsbuffer_block.blockname,
  4     dbsbuffer_location.pnn,
  5     dbsbuffer_dataset.path,
  6     dbsbuffer_dataset_subscription.site,
  7     dbsbuffer_block.create_time
  8  FROM dbsbuffer_dataset_subscription
  9  INNER JOIN dbsbuffer_dataset ON
 10     dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 11  INNER JOIN dbsbuffer_block ON
 12     dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 13  INNER JOIN dbsbuffer_file ON
 14     dbsbuffer_file.block_id = dbsbuffer_block.id
 15  INNER JOIN dbsbuffer_workflow ON
 16     dbsbuffer_workflow.id = dbsbuffer_file.workflow
 17  INNER JOIN dbsbuffer_location ON
 18     dbsbuffer_location.id = dbsbuffer_block.location
 19  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 20  AND dbsbuffer_dataset_subscription.subscribed = 1
 21  AND dbsbuffer_block.status = 'Closed'
 22  AND dbsbuffer_block.deleted = 0
 23  AND dbsbuffer_block.blockname = '/ZeroBias/Run2022C-PromptReco-v1/AOD#d54f99e8-9d28-4567-921f-edce2a207f00'
 24   GROUP BY dbsbuffer_block.blockname,
 25     dbsbuffer_location.pnn,
 26     dbsbuffer_dataset.path,
 27     dbsbuffer_dataset_subscription.site,
 28     dbsbuffer_block.create_time
 29  HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed) ;

no rows selected
```   

It returns no rows as expected, because there are still workflows associated with this block which are not  completed.

Reply:
I am afraid this same faulty logic applies for the extra granularity added by the multiple sites as  well as with the multiple associated workflows